Automated bugfix releases

phargogh opened the pull request:
# Automated Bugfix Releases

This PR sets up automation to perform bugfix releases based on `master` (or whatever the repo default branch is) on a weekly schedule assuming that:

1. The latest commit on `master` does not have any failing tests.
2. The latest commit on `master` has not already been tagged (so, nothing has been merged into `master` since the last bugfix release).
3. The most recent tag on `master` conforms to `^major.minor.bugfix$` notation.  This automation cannot currently special cases like alpha/beta/release candidate tags.
3. The last bugfix release PR is not still in progress.

## What happens weekly
Once a week, a GitHub Actions workflow will fire up to check for the above conditions.  If the conditions are all satisfied, the following actions are taken:

1. The new bugfix version string is determined
2. `Makefile` is updated to reflect the latest commit hash on `master` in the User's Guide
3. `HISTORY.rst` is updated with today's date, and the 'Unreleased Changes' section is commented out in RST.
4. Both the above file changes are committed and pushed to a new branch: `autorelease/$BUGFIX_VERSION`.  (NB: branch restrictions on the repository have been relaxed to allow pushes to `autorelease/*`)  This new branch gives us an easy way to discard the release if needed.
5. A new release object is created for the new bugfix version on the `autorelease/$BUGFIX_VERSION` branch that was just created.  This also creates the `$BUGFIX_VERSION` tag on the branch. ([example](https://github.com/phargogh/invest/releases/tag/3.8.14))
6. A new issue is created in the issue tracker with a reminder to submit the built python packages to PyPI. ([example](https://github.com/phargogh/invest/issues/38))
7. A new Pull Request is opened between `autorelease/$BUGFIX_VERSION` and `master` that includes the changes and a pile of explanatory text about what to do when things go wrong. ([example](https://github.com/phargogh/invest/pull/39))

### On bugfix release errors
While I've taken care to handle most of the kind of ambiguous errors I encountered when developing this workflow, it's very possible that some other errors will arise.  If any part of the release logic above fails, a slack message will be issued to our slack workspace ([example](https://stanford-natcap.slack.com/archives/CUNTH5DPF/p1587153748000100)) to notify us of the failure.

Errors when building the python/binary distributions are not reported in the same way.

### On bugfix release success
If a new bugfix release is created, we'll be notified in the slack channel as well, only the notifications will be in the form of 1) commits pushed to `autorelease/$BUGFIX_VERSION` 2) the `$BUGFIX_VERSION` release created, and 3) the new issue about posting to PyPI.

In response to the release object being created, 2 new build jobs are triggered:

1. Python distributions (a source distribution as well as various platform-specific wheels) are built and uploaded to the new release object.
2. Windows and Mac application binaries are built and uploaded to the new release object.

All of the above builds also archive their own artifacts AND ALSO upload to the releases bucket, so if an upload to the release happens to fail, there are multiple ways to retrieve the artifacts.

Tests are not retriggered by a release.

## Required setup

There are several new pieces of setup to allow this automation to take place, all of which have already been enacted on `natcap/invest`:

1. Branch restrictions must be relaxed to allow pushes to `autorelease/*`.
2. A new secret must be created, `USER_AUTH_TOKEN` with the `public_repo` OAuth scope to enable the release creation to trigger workflows.

## Relevant changes to existing InVEST components

While working on this, it became necessary to make a few improvements to some of our build infrastructure.  These include:

1. The binary workflow has been rejiggered a bit:
    1. Sample data zipfiles are now generated by a separate workflow job to help cut down build time in the Windows build.
    2. Binary artifacts are uploaded to the release only after the binary jobs are finished.
2. A new `Makefile` recipe has been added to nicely extract specific parameter values.
3. `make deploy` will now deploy whatever build targets are found in `dist/`, rather than some OS-specific configuration of files.

davemfish commented on .github/workflows/binary-applications.yml:
```diff
@@ -33,7 +35,9 @@ jobs:
+            - run: git fetch origin +refs/tags/*:refs/tags/*
 
             - name: Fetch git tags
               run: git fetch origin +refs/tags/*:refs/tags/*
```
Is this a redundant git fetch?

phargogh replied:
So it is!  Patched in 333a08bb

davemfish commented on .github/workflows/binary-applications.yml:
```diff
@@ -68,11 +72,21 @@ jobs:
+                  # Only clone the files ... deliberately not cloning LFS data
+                  # Installer needs .invest.json files only, and these are not
+                  # stored in LFS.
+                  GIT_LFS_SKIP_SMUDGE=1 git clone $GIT_URL $LOCALPATH
```
nice!

davemfish commented on .github/workflows/binary-applications.yml:
```diff
@@ -140,7 +160,9 @@ jobs:
+            - run: git fetch origin +refs/tags/*:refs/tags/*
 
             - name: Fetch git tags
               run: git fetch origin +refs/tags/*:refs/tags/*
```
Same as previous job, an extra git fetch?

davemfish replied:
I see this pattern repeated elsewhere too, so probably there is a reason? I'll stop commenting on every occurrence.

phargogh replied:
Hmm this duplication is either sloppy copy-pasting on my part or a merge is putting that line in there.  Hopefully it's the former.  In any case, patching this in 61e7b794.

davemfish commented on ci/release/bugfix-autorelease-branch-pr-body.md:
```diff
@@ -0,0 +1,50 @@
+   Re-tagging and pushing the files to `$TARGET_BRANCH` will cause the release
+   binaries to be rebuilt and re-uploaded to the release object.
+5. Submit a PR from `$GITHUB_REPOSITORY:$TARGET_BRANCH` into `$GITHUB_REPOSITORY:$SOURCE_BRANCH`.
+
```
The PR mentioned in step 5 here has the same target and source as the "Auto PR" that this documetation is describing, right? But the automation only happens on the weekly schedule, so nothing in the prior 4 steps will re-trigger this automated PR. Do I have that right?

phargogh replied:
Yes, that sounds right!  The automation in this file will only be triggered on a schedule and nothing in the previous 4 steps will trigger another PR given what's included in the workflows at present.

davemfish commented on ci/release/bugfix-autorelease-branch-pr-body.md:
```diff
@@ -0,0 +1,50 @@
+3. Take a look at the release object and make sure everything is there:
+   https://github.com/$GITHUB_REPOSITORY/releases/tag/$BUGFIX_VERSION.
+4. Upload python wheels to PyPI.  An issue was created for this,
+   so remember to mark progress there as well.
```
Are there any other important steps? I'm thinking about our normal release checklist. In particular, updating natcap website links (though I think they may be automated somewhere else) and running `invest_autotest.py`. That sort of manual testing starts to defeat the purpose though, and maybe it really is overkill for bugfix releases.

phargogh replied:
I've added a couple of notes in 10a7af5e, including a link to the release checklist which I will need to update to account for these bugfix release-related changes.

Ultimately, I'd love to see all of these forms of tests be automated so that they're just a part of our regular workflow.  In the meantime, running `invest-autotest.py` might be overkill for a bugfix release, and having a quicker release cycle should mean that even if we do have a breaking bug in a bugfix release, we can still issue a fix pretty quickly.

davemfish commented on ci/release/update-history.py:
```diff
@@ -0,0 +1,100 @@
+            'To fix this, uncomment the "Unreleased Changes" section, '
+            'add in whatever notes should have been added before, '
+            'and PR these changes back into master.'
+        )
```
What if the reason that there are commits on master but no changes in HISTORY is because the commits on master are not user-facing (e.g. a PR such as this we're reviewing)? In that case we probably don't want to do the bugfix release because the commits on master are not user-facing, but we also don't need to follow the steps described in this error message. Granted, it's more likely that a HISTORY note was forgetten.

phargogh replied:
Good point! I'd rather have a more accurate error message and then leave it up to the programmer to figure out why things went wrong.  Patched in 001f2b32.

davemfish commented on .github/workflows/auto-release-bugfix.yml:
```diff
@@ -0,0 +1,173 @@
+                      python ci/release/increment-userguide-revision.py
+
+                      # Update HISTORY with the tag and today's date.
+                      python ci/release/update-history.py $BUGFIX_VERSION $(date "+%Y-%m-%d")
```
I'm not so experienced yet with Actions. Does this whole job stop if there's an exception in one of these python scripts? I think we would want it to.

phargogh replied:
Yes, any nonzero exit code will cause the step to immediately abort with an error unless `bash` with `set -e` is used (which we aren't using).

dcdenu4 commented on .github/workflows/auto-release-bugfix.yml:
```diff
@@ -0,0 +1,173 @@
+                      # Python wheels are created in response to the release.
+                      # Create an issue for uploading the wheels to PyPI here
+                      # so that we remember to take care of this.
+                      hub issue create \
+                          --message "Release natcap.invest $BUGFIX_VERSION on PyPI" \
+                          --message "When built, remember to upload wheels for $BUGFIX_VERSION to PyPI." \
+                          --message "Wheels will be uploaded to $(cat $RELEASE_URL_FILE)" \
+                          --labels task
```
Cool!

phargogh replied:
The `hub` CLI is pretty great for avoiding a lot of the heavy lifting with GitHub's various APIs!

phargogh commented:
Thanks for taking a look @davemfish @dcdenu4 !  I've pushed a couple changes and responded to your comments.

@dcdenu4, the PR description isn't in a wiki somewhere, but I've created #94 to make sure I remember to update the [Release Checklist](https://github.com/natcap/invest/wiki/Release-Checklist) (which is the closest related content in the wiki).  Is that along the lines of what you were thinking?

dcdenu4 commented:
@phargogh yea that sounds great. It's just a really nice documentation of what's going on that I might want to refer to! Mostly in the case of me pondering: how do bugfix releases work again? Oh yea, like this!

davemfish commented on .github/workflows/build-py-dists.yml:
```diff
@@ -26,9 +29,7 @@ jobs:
+            - run: git fetch origin +refs/tags/*:refs/tags/*
 
             - name: Fetch git tags
               run: git fetch origin +refs/tags/*:refs/tags/*
```
another duplicate git fetch

phargogh replied:
Patched in fc776a3b

davemfish commented on .github/workflows/build-py-dists.yml:
```diff
@@ -50,14 +54,26 @@ jobs:
+            - run: git fetch origin +refs/tags/*:refs/tags/*
 
             - name: Fetch git tags
               run: git fetch origin +refs/tags/*:refs/tags/*
```
another git fetch

phargogh replied:
Patched in fc776a3b

phargogh commented:
Alright @davemfish , @dcdenu4 , I've fixed the duplicated `git fetch` for tags, and rewritten the Release Checklist (https://github.com/natcap/invest/wiki/Release-Checklist) for the new release processes.  Since tasks related to the release have changed dramatically, I updated the whole document based on our new branching structure.  The old checklist is still there (in the bottom-most section) for easy comparison.

dcdenu4 commented:
Great James! Thanks for updating that checklist.

phargogh commented:
Hey @davemfish I fogot that you had some changes requested!  Sorry about that.  I've addressed them as well as a nontrivial merge conflict on the binary build script.
